// bpmn-js-properties-panel/lib/provider/bpmn/parts/TaskProps.js
'use strict';

var actionEntryFactory = require('bpmn-js-properties-panel/lib/provider/bpmn/parts/implementation/Action'),
    selectEntryFactory = require('bpmn-js-properties-panel/lib/provider/bpmn/parts/implementation/Dropdown'),
    labelEntryFactory = require('bpmn-js-properties-panel/lib/provider/bpmn/parts/implementation/Label'),
    checkboxEntryFactory = require('bpmn-js-properties-panel/lib/provider/bpmn/parts/implementation/CheckBoxEventDefinition'),
    is = require('bpmn-js/lib/util/ModelUtil').is,
    domQuery = require('min-dom').query,
    domQueryAll = require('min-dom').queryAll,
    getBusinessObject = require('bpmn-js/lib/util/ModelUtil').getBusinessObject;
var cmdHelper = require('../../../helper/CmdHelper');
const SUITABILITY_PROCESS_PRICE = "PROCESS_PRICE",
      PP_ADJUSTMENT_OPERATION = "ppAdjustmentOperation",
      PP_ADJUSTMENT_VALUE = "ppAdjustmentValue",
      PP_ROUNDOFF_OPERATION = "ppRoundoffOperation",
      PP_ROUNDOFF_VALUE = "ppRoundoffValue";

module.exports = function(group, element, bpmnFactory, canvas, translate, additionalParams) {
  var bo = getBusinessObject(element);

  function ensureNotNull(prop) {
    if (!prop) {
      throw new Error(prop + ' must be set.');
    }

    return prop;
  }

  function customGet(element, modelProperty) {
    var bo = getBusinessObject(element),
      res = {},
      prop = ensureNotNull(modelProperty);
    res[prop] = bo.get(prop);

    return res;
  };

  function customSet(element, values, modelProperty) {
    var res = {},
      prop = ensureNotNull(modelProperty);
    if (values[prop] !== '') {
      res[prop] = values[prop];
    } else {
      res[prop] = undefined;
    }
    return res;
  };

  function getAdjustmentsProps(element) {
    var res = customGet(element, PP_ADJUSTMENT_OPERATION);
    enableDisableInputs(PP_ADJUSTMENT_VALUE, canDisable(res[PP_ADJUSTMENT_OPERATION]));
    return res;
  };

  function setAdjustmentsProps(element, values) {
    var bo = getBusinessObject(element);
    var conditionEl = domQueryAll("select[name='" + PP_ADJUSTMENT_OPERATION + "']");
    if (conditionEl.length && !conditionEl[0].selectedIndex) {
      bo.ppAdjustmentValue = "";
      enableDisableInputs(PP_ADJUSTMENT_VALUE, true); // isDisable = true
    } else {
      enableDisableInputs(PP_ADJUSTMENT_VALUE, false); // isDisable = false
    }
    var res = customSet(element, values, PP_ADJUSTMENT_OPERATION);
    return cmdHelper.updateProperties(element, res);
  };

  function getAdjustmentsValueProps(element) {
    return customGet(element, PP_ADJUSTMENT_VALUE);
  };

  function setAdjustmentsValueProps(element, values) {
    var res = {},
      prop = ensureNotNull(PP_ADJUSTMENT_VALUE);
    if (values[prop] !== '') {
      if (prop == PP_ADJUSTMENT_VALUE) {
        if (values[prop] >= 0 && values[prop] <= 100) {
          res[prop] = (values[prop].includes(".") ? parseFloat(values[prop]).toFixed(2) : values[prop]);
        }
      }
    } else {
      res[prop] = "";
    }
    return cmdHelper.updateProperties(element, res);
  };


  function getRoundoffProps(element) {
    var res = customGet(element, PP_ROUNDOFF_OPERATION);
    enableDisableInputs(PP_ROUNDOFF_VALUE, canDisable(res[PP_ROUNDOFF_OPERATION]));
    return res;
  };

  function setRoundoffProps(element, values) {
    var bo = getBusinessObject(element);
    var conditionEl = domQueryAll("select[name='" + PP_ROUNDOFF_OPERATION + "']");
    if (conditionEl.length && !conditionEl[0].selectedIndex) {
      bo.ppRoundoffValue = "";
      enableDisableInputs(PP_ROUNDOFF_VALUE, true); // isDisable = true
    } else {
      enableDisableInputs(PP_ROUNDOFF_VALUE, false); // isDisable = false
    }
    var res = customSet(element, values, PP_ROUNDOFF_OPERATION);
    return cmdHelper.updateProperties(element, res);
  };

  function getRoundoffValueProps(element) {
    return customGet(element, PP_ROUNDOFF_VALUE);
  };

  function setRoundoffValueProps(element, values) {
    var res = {},
      prop = ensureNotNull(PP_ROUNDOFF_VALUE);
    if (values[prop] !== '') {
      if (prop == PP_ROUNDOFF_VALUE) {
        res[prop] = values[prop].replace('*','0');
        /* if (values[prop] >= 0 && values[prop] <= 0.99) {
          res[prop] = values[prop];
        } */
      }
    } else {
      res[prop] = "";
    }
    return cmdHelper.updateProperties(element, res);
  };

  function enableDisableInputs(queryElName, isDisable) {
    var valueEl = domQueryAll("div[name='" + queryElName + "']")[0];
    if (valueEl) {
      valueEl.setAttribute("contenteditable", "" + !isDisable);
    }
  }

  function canDisable(condition) {
    if (condition) {
      return (condition == "No adjustments" || condition == "No rounding off");
    } else {
      return true;
    }
  }

  if (is(element, 'bpmn:Task') && bo.customElementName == SUITABILITY_PROCESS_PRICE) {
    var options = { id:"ppAdjustment", label: translate('Adjustments')};
    group.entries = group.entries.concat(labelEntryFactory(element, options, translate));

    options = { id: "ppAdjustment_operation", modelProperty: PP_ADJUSTMENT_OPERATION, label: translate('Operation'), inputArr: additionalParams.config.PPAdjustmentOperation, get: getAdjustmentsProps, set: setAdjustmentsProps};
    group.entries = group.entries.concat(selectEntryFactory(element, options, translate));

    options = { id: "ppAdjustment_value", modelProperty: PP_ADJUSTMENT_VALUE, label: translate('Value (Percentage)'), get: getAdjustmentsValueProps, set: setAdjustmentsValueProps};
    group.entries = group.entries.concat(actionEntryFactory(element, options, translate));

    options = { id: "ppRoundoff", label: translate('Round-off')};
    group.entries = group.entries.concat(labelEntryFactory(element, options, translate));

    options = { id: "ppRoundoff_operation", modelProperty: PP_ROUNDOFF_OPERATION, label: translate('Operation'), inputArr: additionalParams.config.PPRoundoffOperation, get: getRoundoffProps, set: setRoundoffProps};
    group.entries = group.entries.concat(selectEntryFactory(element, options, translate));

    options = { id: "ppRoundoff_value", modelProperty: PP_ROUNDOFF_VALUE, label: translate('Type'), description: "Enter the rounding pattern here. Example: *.00, *.49, *.99, *.29", get: getRoundoffValueProps, set: setRoundoffValueProps};
    group.entries = group.entries.concat(actionEntryFactory(element, options, translate));
  }
};
